src/Datadog.Trace.ClrProfiler.Managed/Integrations/Caching/InterceptedMethodCache.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+
+namespace Datadog.Trace.ClrProfiler.Integrations
+{
+    /// <summary>
+    /// Wrapper class for retrieving instrumented methods and caching them.
+    /// Necessary because our profiling implementation treats all non <see cref="ValueType"/> types as equal for method signature purposes.
+    /// Uses the actual types of arguments from the calling method to differentiate between signatures and caches them.
+    /// </summary>
+    /// <typeparam name="TDelegate">The type of delegate being intercepted.</typeparam>
+    internal class InterceptedMethodCache<TDelegate>
+        where TDelegate : Delegate
+    {
+        private readonly ConcurrentDictionary<string, TDelegate> _methodCache = new ConcurrentDictionary<string, TDelegate>();
+
+        internal TDelegate GetInterceptedMethod(
+            Assembly assembly,
+            string owningType,
+            string methodName,
+            Type[] generics,
+            Type[] parameters)
+        {
+            var methodKey = Interception.MethodKey(genericTypes: generics, parameterTypes: parameters);
+
+            if (!_methodCache.TryGetValue(methodKey, out var method))
+            {
+                var type = assembly.GetType(owningType);
+
+                method = Emit.DynamicMethodBuilder<TDelegate>.CreateMethodCallDelegate(
+                    type,
+                    methodName,
+                    methodParameterTypes: parameters,
+                    methodGenericArguments: generics);
+
+                AddToCache(methodKey, method);
+            }
+
+            return method;
+        }
+
+        private void AddToCache(string key, TDelegate method)
+        {
+            _methodCache.AddOrUpdate(key, (k) => method, (k, m) => method);
```

Comment:
would we ever actually need to do an update? when would that occur?

Reply:
Forced by the ConcurrentDictionary signature, the idea is that the ConcurrentDictionary might be updated by another thread by the time this line gets hit, and so it's a very low level built in mechanism to facilitate concurrency. It just gives you the opportunity to do something different if you end up locked and then an item exists with that key in the dictionary.

Reply:
got it, makes sense. I sort of assumed maybe it was concurrency related... do we need to introduce mutexs?!

Reply:
No, ConcurrentDictionary handles the contention for you.

Reply:
yeah, just surprised we aren't using a `AddOrIgnore` type semantic, if it is already written in there why do we want to update?

Reply:
Fixed in latest